src/main/core/worker.rs

```diff
@@ -270,6 +274,106 @@ impl Worker {
         .unwrap();
     }
 
+    pub fn send_packet(src_host: &Host, packet: *mut cshadow::Packet) {
+        assert!(!packet.is_null());
+
+        let current_time = Worker::current_time().unwrap();
+        let round_end_time = Worker::round_end_time().unwrap();
+
+        let is_completed = current_time >= Worker::with(|w| w.shared.sim_end_time).unwrap();
+        let is_bootstrapping =
+            current_time < Worker::with(|w| w.shared.bootstrap_end_time).unwrap();
+
+        if is_completed {
+            // the simulation is over, don't bother
+            return;
+        }
+
+        let src_ip = unsafe { cshadow::packet_getSourceIP(packet) };
+        let dst_ip = unsafe { cshadow::packet_getDestinationIP(packet) };
+        let payload_size = unsafe { cshadow::packet_getPayloadSize(packet) };
+
+        let src_ip: std::net::Ipv4Addr = u32::from_be(src_ip).into();
+        let dst_ip: std::net::Ipv4Addr = u32::from_be(dst_ip).into();
+
+        let dst_host_id = Worker::with(|w| {
+            w.shared
+                .resolve_ip_to_host_id(dst_ip)
+                .expect("No host ID for dest address {dst_ip}")
+        })
+        .unwrap();
+
+        let src_ip = std::net::IpAddr::V4(src_ip);
+        let dst_ip = std::net::IpAddr::V4(dst_ip);
+
+        // check if network reliability forces us to 'drop' the packet
+        let reliability: f64 = Worker::with(|w| w.shared.reliability(src_ip, dst_ip).unwrap())
+            .unwrap()
+            .try_into()
+            .unwrap();
+        let chance: f64 = src_host.with_random_mut(|r| r.gen());
+
+        // don't drop control packets with length 0, otherwise congestion control has problems
```

Comment:
I know this is just ported over from the C code, but I wonder if we should do something about this at some point. For someone using Shadow to model a lossy network, having 0-length-payload packets be 100% reliable doesn't seem right.

Maybe open an issue to revisit and leave a link here? Or update the comment to be more clear about whether this is/isn't actually a bug in shadow?

Reply:
I converted this into an issue, and left a comment in the code.

https://github.com/shadow/shadow/issues/2517